### PR TITLE
Add channel program for taking recursive snapshots based on property

### PR DESCRIPTION
**Pull request by clinta**
### Motivation and Context

This is a contribution of a channel program that takes recursive snapshots as [suggested](https://github.com/zfsonlinux/zfs/issues/8443#issuecomment-466562897) by @kpande in #8443 .

### Description
Creates the directory contrib/zcp where useful community maintained channel programs can kept and starts with a channel program to satisfy the request in #8443.

### How Has This Been Tested?
Successfully ran the channel program.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [X] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [X] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by kpande** [comment]
can we make the property name configurable? that would make this more useful. thank you for this!

**Comment by clinta** [comment]
Added the ability to specify the property as an optional argument
`Usage: zfs program <zpool> autosnap.lua [noop] [<property>] <snapshot>`

**Comment by clinta** [comment]
The `<pool>` parameter is required by `zfs program`, but the channel program itself operates on whatever dataset you specify, just as if you did `zfs snapshot`.

So `zfs program tank autosnap.lua tank/foo/bar@backup` only creates `tank/foo/bar@backup` and `@backup` snapshots of any child datasets of `tank/foo/bar`. It does not touch anything else in the `tank` pool.

**Comment by kpande** [comment]
thanks! that makes a lot of sense.

File: contrib/zcp/autosnap.lua [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	property: property to check. [default=com.sun:auto-snapshot]
+	snapshot: root snapshot to create example: tank@backup
+]]
+
+if #argv == 0 then
+	error(usage)
+end
+
+noop = false
+-- 'noop' is a workaround for issue #9056
```
Comment (behlendorf) [inline contrib/zcp/autosnap.lua]: It'd be nice to include the proposed fix, adding the `+` prefix to `optstring` from https://github.com/zfsonlinux/zfs/issues/9056#issuecomment-512610853 so we wouldn't need this workaround.  Applying it only to `zfs program` for now should minimize any potential compatibility problems.

Comment (clinta) [reply contrib/zcp/autosnap.lua]: Should that be included here, or should that fix be it's own PR?

**Comment by clinta** [comment]
@behlendorf thanks to your comments on #9056 I was able to remove the workaround and just document that the program needs to be called with a `--` between the name of the script and the parameters.

This makes the argument processing quite a bit simpler and no longer position dependent.

The usage is now `zfs program <pool> autosnap.lua -- [-n] [-p <property>] <snapshot>`

**Comment by clinta** [comment]
Still struggling to get this to build. Looks like there's something more is need in `contrib/zcp/Makefile.am`.

The TEST builds are failing with `make[5]: *** No rule to make target 'autosnap.lua', needed by 'all-am'.  Stop.`

**Comment by behlendorf** [comment]
@clinta sorry about the `Makefile.am`, for now let's keep it super simple.  This should work.

```diff
-zcpdir = $(datarootdir)/zcp
-nodist_zcp_SCRIPTS = autosnap.lua
+EXTRA_DIST = autosnap.lua
```

**Comment by clinta** [comment]
I'm looking at some of the zcp tests and I think I should be able to add testing for this program.

**Comment by behlendorf** [comment]
Let's go ahead and add this to `contrib/zcp` for the moment, then we can follow up as needed with changes to potentially install it and add some test coverage.

**Pull request merged**